fastlane/lanes/codesign.rb

```diff
@@ -53,100 +47,169 @@
-      readonly: true,
-      app_identifier: ALL_JETPACK_BUNDLE_IDENTIFIERS.map { |id| id.sub(JETPACK_BUNDLE_IDENTIFIER, 'com.jetpack.internal') }
+  # @option [Boolean] readonly (default: true) Whether to only fetch existing certificates and profiles, without generating new ones.
+  #
+  private_lane :jetpack_internal_code_signing do |options|
+    update_code_signing_enterprise(
+      app_identifiers: ALL_JETPACK_BUNDLE_IDENTIFIERS.map { |id| id.sub(JETPACK_BUNDLE_IDENTIFIER, 'com.jetpack.internal') },
+      readonly: options.fetch(:readonly, true)
     )
   end
 
-  # Downloads all the required certificates and profiles (using `match``) for the Jetpack App Store builds
+  # Downloads all the required certificates and profiles (using `match``) for the Jetpack App Store builds.
+  # Optionally, it can create any new necessary certificate or profile.
+  #
+  # @option [Boolean] readonly (default: true) Whether to only fetch existing certificates and profiles, without generating new ones.
   #
-  private_lane :jetpack_appstore_code_signing do
-    match(
-      type: 'appstore',
-      team_id: get_required_env('EXT_EXPORT_TEAM_ID'),
-      readonly: true,
-      app_identifier: ALL_JETPACK_BUNDLE_IDENTIFIERS
+  private_lane :jetpack_appstore_code_signing do |options|
+    update_code_signing_app_store(
+      readonly: options.fetch(:readonly, true),
+      app_identifiers: ALL_JETPACK_BUNDLE_IDENTIFIERS
     )
   end
 end
+
+def update_code_signing_enterprise(readonly:, app_identifiers:)
+  if readonly
+    # In readonly mode, we can use the API key
+    api_key_path = APP_STORE_CONNECT_KEY_PATH
+  else
+    # The Enterprise account APIs do not support authentication via API key.
+    # If we want to modify data (readonly = false) we need to authenticate manually.
+    prompt_user_for_app_store_connect_credentials
+    # We also need to pass no API key path, otherwise Fastlane will give
+    # precedence to that authentication mode.
+    api_key_path = nil
+  end
+
+  update_code_signing(
+    type: 'enterprise',
+    # Enterprise builds belong to the "internal" team
+    team_id: get_required_env('INT_EXPORT_TEAM_ID'),
+    readonly: readonly,
+    app_identifiers: app_identifiers,
+    api_key_path: api_key_path
+  )
+end
+
+def update_code_signing_app_store(readonly:, app_identifiers:)
+  update_code_signing(
+    type: 'appstore',
+    # App Store builds belong to the "external" team
+    team_id: get_required_env('EXT_EXPORT_TEAM_ID'),
+    readonly: readonly,
+    app_identifiers: app_identifiers,
+    api_key_path: APP_STORE_CONNECT_KEY_PATH
+  )
+end
+
+def update_code_signing(type:, team_id:, readonly:, app_identifiers:, api_key_path:)
+  # NOTE: It might be neccessary to add `force: true` alongside `readonly: true` in order to regenerate some provisioning profiles.
+  # If this turns out to be a hard requirement, we should consider updating the method with logic to toggle the two setting based on whether we're fetching or renewing.
```

Comment:
I wonder if the reason you had to call `force` this time was because there was no new devices or entitlements or devices that were added since last time those were updated, so the App Store Connect API (or fastlane, not sure which one of the two) maybe considered "eh, there's nothing to update there, so I won't do a thing" — without taking into account in that decision the fact that the certs expired and that's why we wanted them regenerated and updated. Which is why you need to `force`?

While I could imagine that in the past times you ran this (especially before we tidied things up with our MC self-serve tool and all), every time there had been new devices or similar added since last run and that's what made the `force:` unnecessary?

That's just a theory though… I guess we'll see over time, especially next time we'll need to run this lane (… probably not before a year from now 🙃 ).

Reply:
Fascinating 🤔 That sounds plausible, yes. We'll see in the future, hopefully we'll land back here and update with new findings.